swarm/src/lib.rs

```diff
@@ -500,146 +511,100 @@ where
-                                self.behaviour.addresses_of_peer(&peer_id)
-                            }
-                            dial_opts::Opts::WithPeerIdWithAddresses(
-                                dial_opts::WithPeerIdWithAddresses {
-                                    peer_id,
-                                    mut addresses,
-                                    extend_addresses_through_behaviour,
-                                    ..
-                                },
-                            ) => {
-                                if extend_addresses_through_behaviour {
-                                    addresses.extend(self.behaviour.addresses_of_peer(&peer_id))
-                                }
-                                addresses
-                            }
-                            dial_opts::Opts::WithoutPeerIdWithAddress { .. } => {
-                                unreachable!("Due to outer match.")
-                            }
-                        };
+            #[allow(deprecated)]
+            self.behaviour
+                .inject_dial_failure(peer_id, &e, connection_id);
 
-                        let mut unique_addresses = HashSet::new();
-                        addresses.retain(|addr| {
-                            !self.listened_addrs.values().flatten().any(|a| a == addr)
-                                && unique_addresses.insert(addr.clone())
-                        });
+            return Err(e);
+        }
 
-                        if addresses.is_empty() {
-                            let error = DialError::NoAddresses;
-                            #[allow(deprecated)]
-                            self.behaviour
-                                .inject_dial_failure(Some(peer_id), handler, &error);
-                            return Err(error);
-                        };
+        if let Some(peer_id) = peer_id {
+            // Check if peer is banned.
+            if self.banned_peers.contains(&peer_id) {
+                let error = DialError::Banned;
+                #[allow(deprecated)]
+                self.behaviour
+                    .inject_dial_failure(Some(peer_id), &error, connection_id);
+                return Err(error);
+            }
+        }
 
-                        addresses
-                    };
+        let addresses = {
+            let mut addresses_from_opts = dial_opts.get_addresses();
 
-                    (
-                        Some(peer_id),
-                        Either::Left(addresses.into_iter()),
-                        dial_concurrency_factor_override,
-                        role_override,
-                    )
+            match self.behaviour.handle_pending_outbound_connection(
+                peer_id,
+                addresses_from_opts.as_slice(),
+                dial_opts.role_override(),
+                connection_id,
+            ) {
```

Comment:
Just calling this out to make sure I am not missing something. I don't think this is problematic.

Addresses collected through `handle_pending_outbound_connection` might not actually be used in the dial in case the dialer did not enable `extend_addresses_through_behaviour`.

Reply:
Yep, I think this might warrant a `debug` log.

Reply:
> Could we have `handle_pending_outbound_connection` return something like `impl Iterator<Item = &Multiaddr>` instead? That way we'd avoid unnecessary cloning of addresses if they end up not being used anyway.

I'll explore it, good idea!

Reply:
> That way we'd avoid unnecessary cloning of addresses if they end up not being used anyway.

My initial comment was not aimed at the potential performance impact, but rather at the surprise for an implementer of the API, expecting their addresses to be used in the final dial. Instead of optimizing away the allocations, I suggest adding a comment to the trait's method to prevent implementers being surprised.

Reply:
Are we happy with the log statement that I added?

Reply:
Happy!